File: server/services/lease.js
import {poolConnection} from "../database/connection.js"

export async function registerLease(
    userId,
    nationalId,
    dateOfBirth,
    occupation,
    periodEmployedInMonths,
    employerName,
    salary,
    businessAddress,
    phoneNumber,
    currentHomeAddress,
    homePhoneNumber,
    familySize,
    nextOfKin,
    nextOfKinPhoneNumber,
    nextOfKinAddress,
    signature
){
    try {
        //check if lease is registered
        const [leaseCheck] = await poolConnection.query(
            `
            SELECT * FROM leases 
            WHERE user_id = ?;
            `, [userId]
        );

        if (leaseCheck.length < 1) {
            //register lease
            const [lease] = await poolConnection.query(
                `
                INSERT INTO leases (
                    user_id,
                    national_id,
                    dob,
                    occupation,
                    periodEmployedInMonths,
                    employerName,
                    salary,
                    businessAddress,
                    phoneNumber,
                    currentHomeAddress,
                    homePhoneNumber,
                    familySize,
                    nextOfKin,
                    nextOfKinPhoneNumber,
                    nextOfKinAddress,
                    signature
                ) 
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                `, [
                    userId,
                    nationalId,
                    dateOfBirth,
                    occupation,
                    periodEmployedInMonths,
                    employerName,
                    salary,
                    businessAddress,
                    phoneNumber,
                    currentHomeAddress,
                    homePhoneNumber,
                    familySize,
                    nextOfKin,
                    nextOfKinPhoneNumber,
                    nextOfKinAddress,
                    signature
                ]
            );
            
            return {
                id: lease.insertId,
                userId: userId,
                nationalId: nationalId,
                dateOfBirth: dateOfBirth,
                occupation: occupation,
                periodEmployedInMonths: periodEmployedInMonths,
                employerName: employerName,
                salary: salary,
                businessAddress: businessAddress,
                phoneNumber: phoneNumber,
                currentHomeAddress: currentHomeAddress,
                homePhoneNumber: homePhoneNumber,
                familySize: familySize,
                nextOfKin: nextOfKin,
                nextOfKinPhoneNumber: nextOfKinPhoneNumber,
                nextOfKinAddress: nextOfKinAddress,
                signature: signature
            };
        }

        return "Lease already registered"
        
    } catch (error) {
        return error;
    }
}

export async function updateLease(
    leaseId,
    occupation,
    periodEmployedInMonths,
    employerName,
    salary,
    businessAddress,
    phoneNumber,
    currentHomeAddress,
    homePhoneNumber,
    familySize,
    nextOfKin,
    nextOfKinPhoneNumber,
    nextOfKinAddress
){
    try {
        //check if lease is registered
        const lease = await getLease(leaseId);

        if (lease != null) {
            //register lease
            const [leaseQuery] = await poolConnection.query(
                `
                UPDATE leases
                SET occupation = ?, 
                    periodEmployedInMonths = ?,
                    employerName = ?,
                    salary = ?,
                    businessAddress = ?,
                    phoneNumber = ?,
                    currentHomeAddress = ?,
                    homePhoneNumber = ?,
                    familySize = ?,
                    nextOfKin = ?,
                    nextOfKinPhoneNumber = ?,
                    nextOfKinAddress = ?
                WHERE id = ${leaseId};
                `, [
                    occupation,
                    periodEmployedInMonths,
                    employerName,
                    salary,
                    businessAddress,
                    phoneNumber,
                    currentHomeAddress,
                    homePhoneNumber,
                    familySize,
                    nextOfKin,
                    nextOfKinPhoneNumber,
                    nextOfKinAddress
                ]
            );

            const updatedLease = await getLease(leaseId);
            
            return updatedLease;
        }

        return "Lease not found"
        
    } catch (error) {
        return error;
    }
}

export async function getLeases() {
    const [leases] = await poolConnection.query("SELECT * FROM leases")
    return leases
}

export async function getLease(id) {
    const [lease] = await poolConnection.query(
        `
            SELECT * FROM leases
            WHERE id = ?
        `, 
        [id]
        )

    if (lease.length < 1) {
        return "Lease not Found";
    }

    return lease[0]
}

export async function getUserLease(id) {
    const [lease] = await poolConnection.query(
        `
            SELECT * FROM leases
            WHERE user_id = ?
        `, 
        [id]
        )

    if (lease.length < 1) {
        return "Lease not Found";
    }

    return lease[0]
}

File: lib/models/lease.dart
import 'package:equatable/equatable.dart';

class Lease extends Equatable {
  final int id;
  final int userId;
  final String nationalId;
  final String dateOfBirth;
  final String occupation;
  final int periodEmployedInMonths;
  final String employerName;
  final int salary;
  final String businessAddress;
  final String phoneNumber;
  final String currentHomeAddress;
  final String homePhoneNumber;
  final int familySize;
  final String nextOfKin;
  final String nextOfKinPhoneNumber;
  final String nextOfKinAddress;
  final String signature;

  const Lease({
    required this.id,
    required this.userId,
    required this.nationalId,
    required this.dateOfBirth,
    required this.occupation,
    required this.periodEmployedInMonths,
    required this.employerName,
    required this.salary,
    required this.businessAddress,
    required this.phoneNumber,
    required this.currentHomeAddress,
    required this.homePhoneNumber,
    required this.familySize,
    required this.nextOfKin,
    required this.nextOfKinPhoneNumber,
    required this.nextOfKinAddress,
    required this.signature,
  });

  Lease.fromJson(Map<String, dynamic> json)
      : id = json["id"] as int,
        userId = json["user_id"] as int,
        nationalId = json["national_id"] as String,
        dateOfBirth = json["dob"] as String,
        occupation = json["occupation"] as String,
        periodEmployedInMonths = json["periodEmployedInMonths"] as int,
        employerName = json["employerName"] as String,
        salary = json["salary"] as int,
        businessAddress = json["businessAddress"] as String,
        phoneNumber = json["phoneNumber"] as String,
        currentHomeAddress = json["currentHomeAddress"] as String,
        homePhoneNumber = json["homePhoneNumber"] as String,
        familySize = json["familySize"] as int,
        nextOfKin = json["nextOfKin"] as String,
        nextOfKinPhoneNumber = json["nextOfKinPhoneNumber"] as String,
        nextOfKinAddress = json["nextOfKinAddress"] as String,
        signature = json["signature"] as String;

  @override
  List<Object?> get props => [
        id,
        userId,
        nationalId,
        dateOfBirth,
        occupation,
        periodEmployedInMonths,
        employerName,
        salary,
        businessAddress,
        phoneNumber,
        currentHomeAddress,
        homePhoneNumber,
        familySize,
        nextOfKin,
        nextOfKinPhoneNumber,
        nextOfKinAddress,
        signature,
      ];

  @override
  bool? get stringify => true;
}
